### ISSUE TITLE
Test suite failures with Perl 5.10.1

### ISSUE THREAD

**Issue by pghmcfc** [marioroy/mce-shared#3]
Hi Mario, I see these fails with Perl 5.10.1 on Fedora 13 and CentOS 6:
```
$ make test
PERL_DL_NONLAZY=1 /usr/bin/perl "-MExtUtils::Command::MM" "-e" "test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/05_mce_hobo.t ............ ok
t/07_shared_array.t ........ ok
#   Failed test 'shared cache, check unicode keys'
#   at t/07_shared_cache.t line 557.
Wide character in print at /usr/share/perl5/vendor_perl/Test2/Formatter/TAP.pm line 125, <$__ANONIO__> line 191.
#          got: 'á¼Î¸á½±Î½Î±Ï
                             á¿³'
#     expected: 'ἀθάνατῳ'
#   Failed test 'shared cache, check unicode vals'
#   at t/07_shared_cache.t line 558.
Wide character in print at /usr/share/perl5/vendor_perl/Test2/Formatter/TAP.pm line 125, <$__ANONIO__> line 191.
#          got: undef
#     expected: 'ἀλλά τυίδ᾽ ἔλθ᾽, αἴποτα κἀτέρωτα
#    τᾶσ ἔμασ αύδωσ αἴοισα πήλγι
#    ἔκλυεσ πάτροσ δὲ δόμον λίποισα
#    χρύσιον ἦλθεσ.'
#   Failed test 'shared cache, check unicode find keys =~ match (pairs)'
#   at t/07_shared_cache.t line 39.
#   Failed test 'shared cache, check unicode find values =~ match (pairs)'
#   at t/07_shared_cache.t line 39.
# Looks like you failed 4 tests of 154.
t/07_shared_cache.t ........ 
Dubious, test returned 4 (wstat 1024, 0x400)
Failed 4/154 subtests 
#   Failed test 'basic cull'
#   at t/07_shared_cache_lru.t line 55.
Wide character in print at /usr/share/perl5/vendor_perl/Test2/Formatter/TAP.pm line 125.
#     Structures begin differing at:
#          $got->[2] = 'さあ'
#     $expected->[2] = 'ãã'
#   Failed test 'end delete'
#   at t/07_shared_cache_lru.t line 59.
Wide character in print at /usr/share/perl5/vendor_perl/Test2/Formatter/TAP.pm line 125.
#     Structures begin differing at:
#          $got->[2] = 'さあ'
#     $expected->[2] = 'ãã'
#   Failed test 'front delete'
#   at t/07_shared_cache_lru.t line 63.
Wide character in print at /usr/share/perl5/vendor_perl/Test2/Formatter/TAP.pm line 125.
#     Structures begin differing at:
#          $got->[1] = 'さあ'
#     $expected->[1] = 'ãã'
#   Failed test at t/07_shared_cache_lru.t line 120.
Wide character in print at /usr/share/perl5/vendor_perl/Test2/Formatter/TAP.pm line 125, <$__ANONIO__> line 6.
#     Structures begin differing at:
#          $got->[2] = 'さあ'
#     $expected->[2] = 'ãã'
#   Failed test at t/07_shared_cache_lru.t line 124.
Wide character in print at /usr/share/perl5/vendor_perl/Test2/Formatter/TAP.pm line 125, <$__ANONIO__> line 9.
#     Structures begin differing at:
#          $got->[0] = 'さあ'
#     $expected->[0] = 'ãã'
# Looks like you failed 5 tests of 42.
t/07_shared_cache_lru.t .... 
Dubious, test returned 5 (wstat 1280, 0x500)
Failed 5/42 subtests 
t/07_shared_condvar.t ...... ok
t/07_shared_handle.t ....... ok
t/07_shared_handle_utf8.t .. ok
t/07_shared_hash.t ......... ok
t/07_shared_minidb.t ....... ok
t/07_shared_ordhash.t ...... ok
t/07_shared_queue_norm.t ... ok
t/07_shared_queue_prio.t ... ok
t/07_shared_scalar.t ....... ok
t/07_shared_sequence.t ..... ok
Test Summary Report
-------------------
t/07_shared_cache.t      (Wstat: 1024 Tests: 154 Failed: 4)
  Failed tests:  149-152
  Non-zero exit status: 4
t/07_shared_cache_lru.t  (Wstat: 1280 Tests: 42 Failed: 5)
  Failed tests:  8-10, 34, 37
  Non-zero exit status: 5
Files=14, Tests=1097, 13 wallclock secs ( 0.09 usr  0.01 sys +  1.00 cusr  0.25 csys =  1.35 CPU)
Result: FAIL
Failed 2/14 test programs. 9/1097 subtests failed.
make: *** [test_dynamic] Error 255
```
`Sereal` is not installed.
Let me know if there's any other information I can provide that might be useful.

**Comment by marioroy** [comment]
Hi Paul,

Hello and hoping you are well and safe. This issue is unusual. Perl 5.10.1
smoke testing is passing on meta::cpan. Ditto with Strawberry Perl 5.10.1
on Windows. I disabled testing for "unicode" keys. At first glance it looks
to be dualvar not preserving unicode using RedHat's Perl 5.10.1 package.

1.869 Mon May 11 11:00:00 EST 2020 (upcoming release, awaiting further
testing later on today)

  * MCE::Shared::Cache: Disabled testing for "unicode" keys.
    I'm not sure why failing on RedHat 6 using vendor's Perl 5.10.1.
    Smoke test results for Perl 5.10.1 are successful on meta::cpan.
    Keys are stored with expiration time using a dualvar variable.
    Thank you, Paul Howarth for the report.

I want to be sure before making a release and do a couple more checks.
(i.e. construct a non-shared MCE::Shared::Cache object to factor out IPC).
If the problem persists, then likely dualvar not preserving UTF-8.

From 1.868 onwards, everything is serialized in and out of IPC. Numbers are
preserved (do not become strings). UTF-8 strings are preserved included.
Both Storable and Sereal pass. For this reason thought odd seeing tests
failing and wondering if dualvar is broken in the Perl 5.10.1 RPM.

I will do more testing after work hours before making a release.

Regards,
  Mario

**Comment by pghmcfc** [comment]
It might not be perl itself, it could be a module version - lots of modules on EL-6 will be very old now. I can check the versions if there are any you think might be involved.

**Comment by marioroy** [comment]
Hi Paul,

It is so nice to hear from you. The issue on RedHat 6.x boggles my mind. I
spun up the Live CentOS 6.9 DVD in a VM. I factored out MCE::Shared::Server
entirely (no IPC, no Storable, no Sereal). Instead, constructed a
non-shared Cache. The issue persists.

It could be dualvar or the utf8 module. But not sure. So am disabling UTF-8
tests for MCE::Shared::Cache which involves dualvar ( expiration, key ).
All unicode tests pass otherwise for the other modules. Interestingly,
MCE::Shared::Cache 1.868 is passing on Windows 10 - Strawberry Perl 5.10.1,
also smoke testing on meta::cpan including Perl 5.10.1.

1.869 Mon May 11 11:00:00 EST 2020 (upcoming release)

  * Disabled unicode testing for MCE::Shared::Cache on Perl 5.10.1.
    Testing for keys containing unicode is failing on RedHat 6.x,
    but passing in meta::cpan (smoke tests). Thank you, Paul Howarth.

  * Bumped MCE dependency to 1.868.


Regards,
Mario

**Comment by marioroy** [comment]
Hi Paul,

MCE::Shared 1.869 is released. I ended up disabling unicode tests for
MCE::Shared::Cache on Perl 5.10.1. I factored out Storable, Sereal, and IPC
by constructing a non-shared MCE::Shared::Cache->new object. The issue
persists.

It appears that RedHat 6.x will be with us for a while longer.
https://access.redhat.com/support/policy/updates/errata/

Okay, I went and pulled down the vanilla Perl 5.10.1 file from meta::cpan
and compiled it under CentOS 6.9 under /dev/shm where there is available
space -- booted using the Live CentOS 6.9 ISO image.

# yum install gcc make patch
# cd /dev/shm
# tar xjf perl-5.10.1.tar.bz2
# cd perl-5.10.1

# ./Configure -Dprefix=/dev/shm/perl5101 -des -Dusethreads
# make -j 2
# make install

Enabled unicode testing in t/07_shared_cache.t by unblocking the if
statement at the end of the test script. The test is failing with vanilla
Perl 5.10.1. So then, I next pulled Perl 5.12.4 from meta::cpan and built
that. Unicode tests in t/07_shared_cache.t are passing for Perl 5.12.4.

That's a releave.

Passing also with vanilla Perl 5.12.0 built on CentOS 6.9.
Ditto, passing too with Perl 5.11.5 and 5.11.2 dev releases.

Failing with Perl 5.11.1
Failing with Perl 5.11.0

The t/07_shared_cache.t test passes with Perl 5.11.2 and higher.
Interestingly, passing with Strawberry Perl 5.10.1 on Windows. Also seeing
5.10.1 passing on meta::cpan (Linux and BSD).

Well, this is all the time I have for this and am okay leaving MCE::Shared
1.869 as is. MCE 1.868 and MCE::Shared 1.869 have come a long way. I
totally missed doing UTF-8 validation over the years. This is now behind
me. Going forward, everything is preserved going through IPC.

It is out of my control fixing Perl 5.10.1 on RedHat 6.x. At the very least
narrowed down to the minimal Perl 5.11.2 version passing.

Remember if testing, to uncomment/unblock the if statement at the end of
the t/07_shared_cache.t test file using MCE::Shared 1.869.


Regards,
Mario

**Comment by marioroy** [comment]
Hi Paul,

I made a minimum test script. Fails with Perl 5.10.1, passing with Perl
5.11.2 and higher. Solution provided too.

use strict;
use warnings;

package My::Pkg;

  use Scalar::Util 'dualvar';

  sub new { bless [], shift; }
  sub set { push @{ $_[0] }, dualvar(-1, $_[1]); }
  sub get { $_[0]->[0]; }

  1;

package main;

  use utf8;

  my $key = "ἀθάνατῳ";
  my $obj = My::Pkg->new();

  $obj->set($key);
  my $ret = $obj->get();

  print +($key eq $ret) ? "pass\n" : "fail\n";


The issue is Scalar::Util. Updating to a more current Scalar::Util resolves
the problem.

Solution for Perl-5.10.1 in RedHat/CentOS 6.x is updating Scalar::Util to a
more recent version.

Regards,
  Mario

**Comment by pghmcfc** [comment]
Thanks Mario, this is now resolved in MCE-Shared 1.871. All is well again :-)
